File: game/src/main/kotlin/xyz/upperlevel/hgame/world/entity/EntityRegistry.kt
package xyz.upperlevel.hgame.world.entity

import com.badlogic.gdx.physics.box2d.Body
import org.apache.logging.log4j.LogManager
import xyz.upperlevel.hgame.input.BehaviourChangePacket
import xyz.upperlevel.hgame.network.Endpoint
import xyz.upperlevel.hgame.network.NetSide
import xyz.upperlevel.hgame.runSync
import xyz.upperlevel.hgame.world.World
import xyz.upperlevel.hgame.world.player.Player
import xyz.upperlevel.hgame.world.player.PlayerJumpPacket
import xyz.upperlevel.hgame.world.sequence.Sequence

class EntityRegistry(val world: World) {
    private val _entities = HashMap<Int, Entity>()

    val entities: Collection<Entity>
        get() = _entities.values

    private val playerCount = 2
    private var localId = 0
    private var localOffset = -1

    private var endpoint: Endpoint? = null
        set(value) {
            localOffset = when (value!!.side) {
                NetSide.MASTER -> 0
                NetSide.SLAVE -> 1
            }
            field = value
        }

    /**
     * List of bodies that are waiting to be removed from this world.
     * We have to do that because we can only destroy the body after physics world step.
     */
    private val pendingDestruction: MutableList<Body> = ArrayList()

    init {
        Sequence.create()
                .repeat({ sendReset() }, 1000)
                .play()
    }

    fun getEntity(entityId: Int): Entity? {
        return _entities[entityId]
    }

    private fun forceSpawn(entity: Entity) {
        // Assigns entity id if not present.
        if (entity.id < 0) entity.id = localId++ * playerCount + localOffset

        // Spawns the entity.
        _entities[entity.id] = entity

        world.events.call(EntitySpawnEvent(entity))

        logger.info("Created entity type=${entity.entityType.id} with id=${entity.id} (side=${endpoint!!.side.name})")
    }

    fun spawn(entity: Entity) {
        if (!entity.spawned) {
            forceSpawn(entity)
            endpoint!!.send(entity.serialize())
        }
    }

    private fun strictDespawn(entity: Entity) {
        if (entity.spawned) {
            _entities.remove(entity.id)
            entity.id = -1
            pendingDestruction.add(entity.body)
        }
    }

    fun despawn(entity: Entity) {
        if (entity.spawned) {
            strictDespawn(entity)
            // Currently no despawn packet, the logic is the same so both endpoint should despawn at the same time.
        }
    }

    fun clearDestroyedBodies() {
        if (pendingDestruction.size > 0) {
            logger.info("${pendingDestruction.size} entities pending destruction will be removed.")
        }
        pendingDestruction.forEach { body ->
            world.physics.destroyBody(body)
        }
        pendingDestruction.clear()
    }

    private fun createResetPacket(entity: Entity): EntityResetPacket {
        val data = HashMap<String, Any>()
        entity.fillResetPacket(data)
        return EntityResetPacket(entity.id, data)
    }

    fun sendReset() {
        logger.debug("sending reset")
        endpoint?.let { endp ->
            _entities.map { it.value }
                    .filter { it.active }
                    .map { createResetPacket(it) }
                    .forEach { endp.send(it) }
        }
    }

    fun initEndpoint(endpoint: Endpoint) {
        this.endpoint = endpoint

        endpoint.events.register(EntitySpawnPacket::class.java, { packet ->
            runSync {
                logger.info("Received entity ${packet.entityTypeId} and id=${packet.entityId}")

                val entity = EntityTypes[packet.entityTypeId]!!.create(world, false)
                entity.deserialize(packet)
                forceSpawn(entity)
            }
        })
        endpoint.events.register(BehaviourChangePacket::class.java, { packet ->
            runSync {
                val res = _entities[packet.actorId]
                        ?.behaviour
                        ?.layers
                        ?.get(packet.layerIndex)
                        ?.active(packet.behaviour)
                if (res == null) {
                    logger.warn("Invalid packet")
                }
            }
        })
        endpoint.events.register(PlayerJumpPacket::class.java, { packet ->
            runSync {
                val player = _entities[packet.entityId]
                if (player == null || player !is Player) {
                    logger.warn("Invalid packet: invalid entity")
                    return@runSync
                }// TODO: log error
                player.jump()
            }
        })
        endpoint.events.register(EntityImpulsePacket::class.java, { packet ->
            runSync {
                _entities[packet.entityId]!!.impulse(packet.powerX, packet.powerY, packet.pointX, packet.pointY, false)
            }
        })
        endpoint.events.register(EntityResetPacket::class.java, {packet ->
            runSync {
                logger.debug("received reset {}", packet.entityId)
                val entity = _entities[packet.entityId]
                if (entity == null) {
                    logger.warn("Invalid reset packet: no entity with id ${packet.entityId}")
                    return@runSync
                }
                entity.onReset(packet.data)
            }
        })
    }

    companion object {
        private val logger = LogManager.getLogger()
    }
}

File: game/src/entity/entityRegistry.ts
import {BehaviourChangePacket, EntityResetPacket, EntitySpawnPacket} from "../protocol";
import {Entity} from "./entity";
import {World} from "../world/world";
import {EntityTypes} from "./entityTypes";

export class EntityRegistry {
    entities = new Map<number, Entity>();

    private playerCount = 0;
    private localOffset = 0;

    private localId = 0;

    private timerId: any;

    readonly world: World;

    constructor(world: World) {
        this.world = world;
    }

    onEnable() {
        this.timerId = window.setInterval(this.sendReset.bind(this), 1000);
    }

    onDisable() {
        window.clearTimeout(this.timerId);
    }

    setup(playerCount: number, playerIndex: number) {
        this.playerCount = playerCount;
        this.localOffset = playerIndex;
    }

    onSpawn(packet: EntitySpawnPacket) {
        let type = EntityTypes.get(packet.entityType);

        if (type == null) {
            console.log("Invalid spawn: entity type " + packet.entityType + " not defined");
            return
        }

        let entity = type.create(this.world, false);
        entity.id = packet.entityId;
        entity.sprite.x = packet.x;
        entity.sprite.y = packet.y;
        entity.sprite.scale.x = packet.isFacingLeft ? -1 : 1;
        if (packet.meta != null) {
            entity.loadSpawnMeta(packet.meta);
        }

        this.forceSpawn(entity);
    }

    onBehaviourChange(packet: BehaviourChangePacket) {
        const entity: any = this.entities.get(packet.actorId);

        // If the entity supports behaviours.
        if (entity.behaviour != null) {
            const layer = entity.behaviour.layers[packet.layerIndex];
            if (layer == null) {
                throw `Layer ${packet.layerIndex} is null for entity: ${packet.actorId}`;
            }

            const behaviour = layer.behaviours.get(packet.behaviour);
            if (behaviour == null) {
                throw `Behaviour ${packet.behaviour} is null for entity: ${packet.actorId}`;
            }

            layer.active = behaviour;
        } else {
            throw `Entity ${packet.actorId} does not support behaviours`
        }
    }

    sendReset() {
        for (let [id, entity] of this.entities) {
            if (!entity.active) continue;
            entity.sendReset();
        }
    }

    onResetPacket(packet: EntityResetPacket) {
        let entity = this.entities.get(packet.entityId);
        if (entity == null) {
            throw `Entity ${packet.entityId} not found`;
        }
        entity.onReset(packet);
    }

    getEntity(id: number): Entity | undefined {
        return this.entities.get(id);
    }

    forceSpawn(entity: Entity) {
        if (entity.id == -1) {
            entity.id = this.localId++ * this.playerCount + this.localOffset;
        }

        this.world.app.stage.addChild(entity.sprite);
        this.entities.set(entity.id, entity);
    }

    spawn(entity: Entity) {
        if (entity.id != -1) return;
        this.forceSpawn(entity);

        let packet: EntitySpawnPacket = {
            type: "entity_spawn",
            entityType: entity.type.id,
            entityId: entity.id,
            x: entity.x,
            y: entity.y,
            isFacingLeft: entity.isFacingLeft,
            meta: entity.createSpawnMeta(),
        };
        entity.world.sendPacket(packet);
    }

    despawn(entity: Entity) {
        if (entity.id == -1) {
            console.warn("Trying to despawn a non-spawned entity!");
            return;
        }
        this.entities.delete(entity.id);
        entity.remove();
    }

    onPrePhysicsStep(timedelta: number) {
        for (let [id, entity] of this.entities) {
            entity.onPrePhysics(timedelta);
        }
    }

    onUpdate(timedelta: number): any {
        for (let [id, entity] of this.entities) {
            entity.onUpdate(timedelta);
        }
    }
}
